### PR TITLE
[stdlib] Create new string sub-package

### LINKED ISSUE #3871: [Feature Request] [proposal] Move everything string-related to a sub-package `/stdlib/collections/string/`
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

Create a sub-package `/collections/string/` that will contain:
- `string.mojo`
- `string_slice.mojo`
- `format.mojo`
- `_unicode*.mojo`
- `_utf8_validation.mojo`
- `inline_string.mojo`

And eventually we can also add other files taking related functionality out to their own file to un-bloat `string.mojo` and `string_slice.mojo`

### What is your motivation for this change?

More things in `/utils/` are string-related than not

### Any other details?

...

### PR DESCRIPTION
**Pull request by martinvuyk**
Closes #3871

### CODE REVIEW

**Comment by JoeLoser** [comment]
@martinvuyk do you mind fixing the conflicts and then I can sync this in?  I'd like to get this in and then have you rebase your other string-related PRs on top of this.  Does that work for you?

File: stdlib/src/utils/__init__.mojo [hunk 1]
```diff
@@ -13,11 +13,11 @@
 """Implements the utils package."""
 
 from .index import Index, IndexList, product
-from .inline_string import InlineString
+from collections.string.inline_string import InlineString
```
Comment (JoeLoser) [inline stdlib/src/utils/__init__.mojo:16]: **Question** Do you plan on removing this re-export from `utils` in favor of doing it in `collections/string/__init__.mojo`?  Then we can fix the imports everywhere.

**Comment by martinvuyk** [comment]
@JoeLoser done

File: stdlib/src/collections/string/string_slice.mojo [hunk 2]
```diff
@@ -12,27 +12,25 @@
...
+from collections.string._utf8_validation import _is_valid_utf8
+from collections.string.string import _atof, _atol, _isspace
 from memory import UnsafePointer, memcmp, memcpy, Span
 from memory.memory import _memcmp_impl_unconstrained
-
-from utils.format import _CurlyEntryFormattable, _FormatCurlyEntry
-
-from ._utf8_validation import _is_valid_utf8
+from sys import bitwidthof, simdwidthof
+from sys.intrinsics import unlikely, likely
```
Comment (JoeLoser) [inline stdlib/src/collections/string/string_slice.mojo:32]: FYI I have a PR up internally to re-export `likely`, `unlikely`, etc from the `sys` module.

**Review by JoeLoser — APPROVED** [review]

**Comment by JoeLoser** [comment]
!sync

**Comment by JoeLoser** [comment]
FYI this is causing some breakage in tooling-related tests that we'll need to fix up before we can merge internally and keep building on top of.

**Pull request closed without merging**